Sources/HummingbirdCore/Utils/FlatDictionary.swift

```diff
@@ -26,104 +26,154 @@ public struct FlatDictionary<Key: Hashable, Value>: Collection, ExpressibleByDic
     public typealias Element = (key: Key, value: Value)
     public typealias Index = Array<Element>.Index
 
+    @usableFromInline
+    struct Storage {
+        @usableFromInline internal /* private */ var elements: [Element]
+        @usableFromInline internal /* private */ var hashKeys: [Int]
+
+        @usableFromInline
+        init(elements: [Element], hashKeys: [Int]) {
+            self.elements = elements
+            self.hashKeys = hashKeys
+        }
+    }
+
+    @usableFromInline internal /* private */ var storage: Storage?
+
     // MARK: Collection requirements
 
     /// The position of the first element
-    public var startIndex: Index { self.elements.startIndex }
+    @inlinable
+    public var startIndex: Index { self.storage?.elements.startIndex ?? 0 }
     /// The position of the element just after the last element
-    public var endIndex: Index { self.elements.endIndex }
+    @inlinable
+    public var endIndex: Index { self.storage?.elements.endIndex ?? 0 }
     /// Access element at specific position
-    public subscript(_ index: Index) -> Element { return self.elements[index] }
+    @inlinable
+    public subscript(_ index: Index) -> Element { return self.storage!.elements[index] }
```

Comment:
`!` isn't ideal here, perhaps we should add a precondition that `storage` exist and provide an out of range message or something similar, given the index will be out of range